spras/config/algorithms.py

```diff
@@ -0,0 +1,161 @@
+"""
+Dynamic construction of algorithm parameters with runtime type information for
+parameter combinations. This has been isolated from schema.py as it is not declarative,
+and rather mainly contains validators and lower-level pydantic code.
+"""
+import ast
+import copy
+from typing import Annotated, Any, Callable, Literal, Optional, Union, cast, get_args
+
+import numpy as np
+from pydantic import BaseModel, BeforeValidator, ConfigDict, Field, create_model
+
+from spras.runner import algorithms
+
+__all__ = ['AlgorithmUnion']
+
+def is_numpy_friendly(type: type[Any] | None) -> bool:
+    """
+    Whether the passed in type can have any numpy helpers.
+    This is mainly used to provide hints in the JSON schema.
+    """
+    allowed_types = (int, float)
+
+    # check basic types, then check optional types
+    return type in allowed_types or \
+        any([arg for arg in get_args(type) if arg in allowed_types])
+
+def python_evalish_coerce(value: Any) -> Any:
```

Comment:
Could you add a more detailed comment on what is happening in this function and why we are using this. This way someone doesn't need to read the code fully to understand what is happening.